### PR TITLE
Adding ttl_delete parameter to metadata for DynamoDB

### LINKED ISSUE #4560: Add metadata to DynamoDB stream Events to indicate if they are deleted via TTL
**Is your feature request related to a problem? Please describe.**
As a user of the DynamoDB source, I would like to differentiate between items deleted via TTL for DynamoDB streams, vs items that were explicitly deleted. 

DynamoDB streams provides this information in the `userIdentity` field of the stream record (https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_streams_Record.html)

**Describe the solution you'd like**
An additional metadata field in DynamoDB stream events. 

```
deleted_with_ttl: true 
```

If the record was not a result of TTL deletion, this metadata key will not get applied to the Event at all.

**Additional context**

### PR DESCRIPTION
**Pull request by LeeroyHannigan**
### Description
This change allows users to identify the source of a REMOVE event from DynamoDB streams,  if it was a manual delete or a TTL delete.
 
### Issues Resolved
Resolves #4560 
 
### Check List
- [X] New functionality includes testing.
- [ ] New functionality has a documentation issue. Please link to it in this PR.
 - [ ] New functionality has javadoc added
- [X] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/converter/RecordConverter.java [hunk 1]
```diff
@@ -76,14 +77,16 @@ void flushBuffer() throws Exception {
...
      * @throws Exception Exception if failed to write to buffer.
      */
     public void addToBuffer(final AcknowledgementSet acknowledgementSet,
                             final Map<String, Object> data,
                             final Map<String, Object> keys,
                             final long eventCreationTimeMillis,
                             final long eventVersionNumber,
-                            final String eventName) throws Exception {
+                            final String eventName,
+                            final Boolean userIdentity) throws Exception {
```
Comment (graytaylor0) [inline data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/converter/RecordConverter.java:89]: This boolean should named according to whether it's a TTL delete. Alternatively, it may make sense to make this a String and do the boolean conversion to TTL delete in this method

File: data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/converter/RecordConverter.java [hunk 2]
```diff
@@ -102,6 +105,7 @@ public void addToBuffer(final AcknowledgementSet acknowledgementSet,
         eventMetadata.setAttribute(DDB_STREAM_EVENT_NAME_METADATA_ATTRIBUTE, eventName);
         eventMetadata.setAttribute(EVENT_NAME_BULK_ACTION_METADATA_ATTRIBUTE, mapStreamEventNameToBulkAction(eventName));
         eventMetadata.setAttribute(EVENT_VERSION_FROM_TIMESTAMP, eventVersionNumber);
+        eventMetadata.setAttribute(DDB_STREAM_EVENT_USER_IDENTITY, userIdentity);
```
Comment (graytaylor0) [inline data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/converter/RecordConverter.java:108]: It might make sense to only add this metadata attribute when it is a TTL delete. That way in the configuration later on for routes and things of that sort, users can just do a simple check on

```
getMetadata("ttl_delete") != null
```

rather than having to check explicitly for a true value with `getMetadata("ttl_delete") == true` . I am fine with it either way though.

File: data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/converter/StreamRecordConverter.java [hunk 3]
```diff
@@ -80,6 +80,7 @@ public void writeToBuffer(final AcknowledgementSet acknowledgementSet, List<Reco
         int eventCount = 0;
         for (Record record : records) {
             final long bytes = record.dynamodb().sizeBytes();
+            final Boolean userIdentity = record.userIdentity() != null && "dynamodb.amazonaws.com".equals(record.userIdentity().principalId());
```
Comment (graytaylor0) [inline data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/converter/StreamRecordConverter.java:83]: This variable should be named `isTtlDelete`, or the userIdentity could be passed to `addToBuffer` directly as a String like my comment above.

Also, would be good to have "dynamodb.amazonaws.com" as a private static variable whenever it is used.

**Review by graytaylor0 — APPROVED** [review]
Thanks for making this contribution! Made some minor comments that would be nice to address but overall does look good.

File: data-prepper-plugins/dynamodb-source/src/test/java/org/opensearch/dataprepper/plugins/source/dynamodb/converter/StreamRecordConverterTest.java [hunk 4]
```diff
@@ -476,6 +480,39 @@ void remove_record_without_use_old_image_on_delete_uses_new_image() throws Excep
         verify(bytesProcessedSummary).record(record.dynamodb().sizeBytes());
     }
 
+    @Test
+    void test_writeSingleRecordToBuffer_with_userIdentity() throws Exception {
+
+        when(streamConfig.getStreamViewForRemoves()).thenReturn(StreamViewType.OLD_IMAGE);
+
+        final Identity userIdentity = Identity.builder()
+                .principalId("dynamodb.amazonaws.com")
```
Comment (dlvenable) [inline data-prepper-plugins/dynamodb-source/src/test/java/org/opensearch/dataprepper/plugins/source/dynamodb/converter/StreamRecordConverterTest.java:489]: Please include a test case where the userIdentity is not null, but the principal is something other than this known principle.

It would cover the case of `&& "dynamodb.amazonaws.com".equals(record.userIdentity().principalId())`.

**Review by dlvenable — CHANGES_REQUESTED** [review]
Thank you @LeeroyHannigan for this nice feature! I have a few comments, many also shared by @graytaylor0 .

File: data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/converter/StreamRecordConverter.java (hunk 3 above)
Comment (dlvenable) [inline data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/converter/StreamRecordConverter.java:83]: Also, I just checked the documentation it states that `Records[].userIdentity.type` will equal `"Service"`. Let's also check that condition.

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/converter/MetadataKeyAttributes.java [hunk 5]
```diff
@@ -21,4 +21,5 @@ public class MetadataKeyAttributes {
     static final String DDB_STREAM_EVENT_NAME_METADATA_ATTRIBUTE = "dynamodb_event_name";
 
     static final String EVENT_TABLE_NAME_METADATA_ATTRIBUTE = "table_name";
+    static final String DDB_STREAM_EVENT_USER_IDENTITY = "ttl_delete";
```
[the remark this answers is not in the document]
Comment (LeeroyHannigan) [reply data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/converter/MetadataKeyAttributes.java]: Renamed static variable

File: data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/converter/StreamRecordConverter.java (hunk 3 above)
Comment (LeeroyHannigan) [reply data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/converter/StreamRecordConverter.java]: Added the extra check for `userIdentity.type`

Comment (LeeroyHannigan) [reply data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/converter/StreamRecordConverter.java]: I moved the logic to `addToBuffer` but instead of passing a String, I was forced to pass the Identity as it also contains `Type` which was suggested for checking.

**Comment by LeeroyHannigan** [comment]
Thank you both for review. Incorporated both your feedback, let me know if the current version is acceptable.

**Review by graytaylor0 — APPROVED** [review]
Looks good thanks @LeeroyHannigan! Were you able to test this end to end, or should that be an action item for the Data Prepper maintainers?

**Review by dlvenable — APPROVED** [review]
Thank you @LeeroyHannigan for this contribution!

**Comment by LeeroyHannigan** [comment]
Hi Team, I see this was merged, but I don't see it in the documentation. Is that another PR specifically for documentation?

https://opensearch.org/docs/latest/data-prepper/pipelines/configuration/sources/dynamo-db/

**Comment by eduardowillame** [comment]
> Hi Team, I see this was merged, but I don't see it in the documentation. Is that another PR specifically for documentation?
> 
> https://opensearch.org/docs/latest/data-prepper/pipelines/configuration/sources/dynamo-db/

Still wanted to know if there any pending request to update the documentation. I just checked and it is not updated yet.

**Comment by san81** [comment]
Yes, It has to be another PR on this repo https://github.com/opensearch-project/documentation-website/tree/main/_data-prepper/pipelines/configuration to update the documentation.
